File: CocoatainerExample/CocoatainerExample/CocoaPowder.m
//
//  CocoaPowder.m
//  CocoatainerExample
//
//  Created by Jeffrey Bakker on 2015-05-13.
//  Copyright (c) 2015 Jeffrey Bakker. All rights reserved.
//

#import "CocoaPowder.h"

@interface CocoaPowder ()
{
@private
    id<Topping> _topping;
}
@end
@implementation CocoaPowder

-(id)initWithTopping:(id<Topping>)topping
{
    self = [super init];
    if (self)
    {
        _topping = topping;
    }
    return self;
}

-(void)dealloc
{
    NSLog(@"This cocoa powder has coagulated at the bottom.");
}

-(void)start
{
    NSLog(@"Creating %@ mix with %@ topping",
          NSStringFromClass([self class]), [_topping name]);
}

-(void)shovel
{
    NSLog(@"Shovel three tablespoons of mixture.");
}

@end

File: CocoatainerSwiftExample/CocoatainerSwiftExample/SwiftCocoaMug/CocoaPowder.swift
//
//  CocoaPowder.swift
//  CocoatainerSwiftExample
//
//  Created by Jeffrey Bakker on 2015-05-20.
//  Copyright (c) 2015 Jeffrey Bakker. All rights reserved.
//

import Cocoatainer

public class CocoaPowder: NSObject, Mixture, CCTStartable
{
    var topping: Topping

    public init(topping: Topping)
    {
        self.topping = topping
    }

    deinit
    {
        print("This cocoa powder has coagulated at the bottom.")
    }

    public func start()
    {
        print("Creating \(String(describing: type(of: self))) mix with \(topping.name()) topping.")
    }

    public func shovel()
    {
        print("Shovel three tablespoons of mixture.")
    }
}
